use t.TempDir instead of ioutil.TempDir in unit tests

ianwoolf opened the pull request:
Signed-off-by: ianwoolf <btw515wolf2@gmail.com>

Updates #9594

LeviHarrison reviewed (approved):
Thanks! And welcome to the project!

LeviHarrison commented on promql/engine_test.go:
```diff
@@ -39,8 +38,7 @@ func TestMain(m *testing.M) {
-	dir, err := ioutil.TempDir("", "test_concurrency")
-	require.NoError(t, err)
+	dir := t.TempDir()
 	defer os.RemoveAll(dir)
```
Apologies, I thought the `test_windows` failure was a flake (as it usually is), but there is actually a problem. `defer os.RemoveAll(dir)` should be removed as well, as `t.TempDir()` takes care of deleting the files.

ianwoolf replied:
emm...     I didn't notice the the tmp dir was automatically removed.

LeviHarrison commented:
I think there's just one last thing: the `os` import needs to be removed.

```
# github.com/prometheus/prometheus/promql [github.com/prometheus/prometheus/promql.test]
promql/engine_test.go:19:2: imported and not used: "os"
```

ianwoolf commented:
> I think there's just one last thing: the `os` import needs to be removed.
> 
> ```
> # github.com/prometheus/prometheus/promql [github.com/prometheus/prometheus/promql.test]
> promql/engine_test.go:19:2: imported and not used: "os"
> ```

LeviHarrison commented:
OK, so I think I've figured out why `test_windows` keeps failing with the following error: `testing.go:967: TempDir RemoveAll cleanup: remove C:\Users\circleci\AppData\Local\Temp\TestQueryConcurrency3635900781\001\queries.active: The process cannot access the file because it is being used by another process.` This error happens because the query logger is still using the active query file when the temporary directory is removed. That file should be closed before the temporary directory is removed, and we do stop the query engine (which should close the file), just using a `defer` statement. That worked before this change when the temporary directory was also removed using a `defer` statement. But now, `t.TempDir` removes the directory using a [cleanup function](https://pkg.go.dev/testing#T.Cleanup), which is called before the `defer`red functions are. Therefore, the directory is attempted to be removed before the query engine is closed, causing the error.

In order to fix this, these two `defer`red functions should be registered as [cleanup functions](https://pkg.go.dev/testing#T.Cleanup) instead:

https://github.com/ianwoolf/prometheus/blob/4ae36a0575b4a4bc243862231a0cf4e613d7970c/promql/engine_test.go#L53

https://github.com/ianwoolf/prometheus/blob/4ae36a0575b4a4bc243862231a0cf4e613d7970c/promql/engine_test.go#L58
(I don't think this one's actually necessary but we might as well)

That should fix it.

ianwoolf commented:
> OK, so I think I've figured out why `test_windows` keeps failing with the following error: `testing.go:967: TempDir RemoveAll cleanup: remove C:\Users\circleci\AppData\Local\Temp\TestQueryConcurrency3635900781\001\queries.active: The process cannot access the file because it is being used by another process.` This error happens because the query logger is still using the active query file when the temporary directory is removed. That file should be closed before the temporary directory is removed, and we do stop the query engine (which should close the file), just using a `defer` statement. That worked before this change when the temporary directory was also removed using a `defer` statement. But now, `t.TempDir` removes the directory using a [cleanup function](https://pkg.go.dev/testing#T.Cleanup), which is called before the `defer`red functions are. Therefore, the directory is attempted to be removed before the query engine is closed, causing the error.
> 
> In order to fix this, these two `defer`red functions should be registered as [cleanup functions](https://pkg.go.dev/testing#T.Cleanup) instead:
> 
> https://github.com/ianwoolf/prometheus/blob/4ae36a0575b4a4bc243862231a0cf4e613d7970c/promql/engine_test.go#L53
> 
> https://github.com/ianwoolf/prometheus/blob/4ae36a0575b4a4bc243862231a0cf4e613d7970c/promql/engine_test.go#L58 (I don't think this one's actually necessary but we might as well)
> 
> That should fix it.

3ks.   i will work on it.

ianwoolf commented:
> OK, so I think I've figured out why `test_windows` keeps failing with the following error: `testing.go:967: TempDir RemoveAll cleanup: remove C:\Users\circleci\AppData\Local\Temp\TestQueryConcurrency3635900781\001\queries.active: The process cannot access the file because it is being used by another process.` This error happens because the query logger is still using the active query file when the temporary directory is removed. That file should be closed before the temporary directory is removed, and we do stop the query engine (which should close the file), just using a `defer` statement. That worked before this change when the temporary directory was also removed using a `defer` statement. But now, `t.TempDir` removes the directory using a [cleanup function](https://pkg.go.dev/testing#T.Cleanup), which is called before the `defer`red functions are. Therefore, the directory is attempted to be removed before the query engine is closed, causing the error.
> 
> In order to fix this, these two `defer`red functions should be registered as [cleanup functions](https://pkg.go.dev/testing#T.Cleanup) instead:
> 
> https://github.com/ianwoolf/prometheus/blob/4ae36a0575b4a4bc243862231a0cf4e613d7970c/promql/engine_test.go#L53
> 
> https://github.com/ianwoolf/prometheus/blob/4ae36a0575b4a4bc243862231a0cf4e613d7970c/promql/engine_test.go#L58 (I don't think this one's actually necessary but we might as well)
> 
> That should fix it.

intresting...   i updated code,  register the `defer` as `t.CleanUp functioins`. But it didn't work, the windows test fail again.

So i found a windows and just run `go test ./...` under the directory `promql`, the test is fail. But it did pass on my mac. And i check the directory `C:\Users\86136\AppData\Local\Temp\TestQ*` of my windows, i get

    Mode                LastWriteTime         Length Name
    ----                -------------         ------ ----
    d-----        2022/2/27      0:12                TestQueryConcurrency18368351
    d-----        2022/2/27      0:16                TestQueryConcurrency2333167212
    d-----        2022/2/26     23:59                TestQueryConcurrency2363779818
    d-----        2022/2/27      0:44                TestQueryConcurrency3202761245

seems these temp directories were not deleted. i don't know why? did i miss something or it was caused by somethinf of windows?

roidelapluie commented:
cc @ianwoolf can you please rebase this pull request? Thanks!

bboreham commented:
Discussed in the bug scrub; apologies we dropped the ball in July.  This still looks worthwhile, however at the last CI run Windows tests failed and something needs a DCO signed-off line.

@ianwoolf could you add the signed-off, rebase again and we can try to figure out the Windows test?

aknuds1 commented:
I think this PR can now be safely closed, since after rebasing on main and fixing conflicts, there are no changes left.

bboreham commented:
Thanks!
